src/IonBinaryWriter.ts

```diff
@@ -58,678 +58,658 @@ enum States {
+    addChild(child: Node, fieldName?: number[]) : void {
+        if (isNullOrUndefined(fieldName)) {
+            throw new Error("Cannot add a value to a struct without a field name");
+        }
+        this.fields.push({name: fieldName, value: child});
     }
-    this.fields.push({name: fieldName, value: child});
-  }
 
-  getValueLength() : number {
-    let valueLength: number = 0;
-    for (let field of this.fields) {
-      valueLength += field.name.length;
-      valueLength += field.value.getLength();
+    getValueLength() : number {
+        let valueLength: number = 0;
+        for (let field of this.fields) {
+            valueLength += field.name.length;
+            valueLength += field.value.getLength();
+        }
+        return valueLength;
     }
-    return valueLength;
-  }
 
-  getLength() : number {
-    if (isUndefined(this.length)) {
-      this.length = super.getLength();
+    getLength() : number {
+        if (isUndefined(this.length)) {
+            this.length = super.getLength();
+        }
+        return this.length;
     }
-    return this.length;
-  }
 
-  write() : void {
-    this.writeAnnotations();
-    this.writeTypeDescriptorAndLength(this.typeCode, false, this.getValueLength());
-    for (let field of this.fields) {
-      this.writer.writeBytes(field.name);
-      field.value.write();
+    write() : void {
+        this.writeAnnotations();
+        this.writeTypeDescriptorAndLength(this.typeCode, false, this.getValueLength());
+        for (let field of this.fields) {
+            this.writer.writeBytes(field.name);
+            field.value.write();
+        }
     }
-  }
 }
-
+//These feel wrong
 export abstract class LeafNode extends AbstractNode {
-  addChild(child: Node, name?: number[]) : void {
-    throw new Error("Cannot add a child to a leaf node");
-  }
+    addChild(child: Node, name?: number[]) : void {
+        throw new Error("Cannot add a child to a leaf node");
+    }
 
-  isContainer() : boolean {
-    return false;
-  }
+    isContainer() : boolean {
+        return false;
+    }
 }
 
 class BooleanNode extends LeafNode {
-  constructor(writer: LowLevelBinaryWriter, parent: Node, annotations: number[], private readonly value: boolean) {
-    super(writer, parent, TypeCodes.BOOL, annotations);
-  }
-
-  write() : void {
-    this.writeAnnotations();
-    this.writeTypeDescriptorAndLength(this.typeCode, false, this.value ? 1 : 0);
-  }
-
-  getValueLength() : number {
-    return 0;
-  }
+    private readonly value: boolean;
+
+    constructor(writer: LowLevelBinaryWriter, parent: Node, annotations: number[], inputValue : boolean) {
+        super(writer, parent, TypeCodes.BOOL, annotations);
+        this.value = inputValue;
+    }
+
+    write() : void {
+        this.writeAnnotations();
+        this.writeTypeDescriptorAndLength(this.typeCode, false, this.value ? 1 : 0);
+    }
+
+    getValueLength() : number {//TODO wtf
```

Comment:
Boolean values are always encoded in the lower nibble of the type ID byte. So if getValueLength() measures the length past the type ID byte, this is right.